### ISSUE TITLE
Can a NoAho object be pickled ?

### ISSUE THREAD

**Issue by simonatdrg** [JDonner/NoAho#4]
Great module - I've just started to use it to look for terms in Japanese text, and it seems to work well.

Is it possible to pickle the object ?  we  add many (~ 100K ) terms which take a while to load, and would like to persist the object once we have it. Have played with pickle.Pickler but the file created simply has header information.

**Comment by JDonner** [comment]
No, but that's a great idea. I might look into it in a couple of weeks - until then, pull requests are always welcome!

**Comment by geovedi** [comment]
This is my simple workaround for this.

```diff
diff --git a/src/noaho.pyx b/src/noaho.pyx
index 9a13feb..de02014 100644
--- a/src/noaho.pyx
+++ b/src/noaho.pyx
@@ -64,9 +64,11 @@ cdef extern from "array-aho.h":
 cdef class NoAho:
     cdef AhoCorasickTrie *thisptr
     cdef object payloads_to_decref
+    cdef dict dictionary
     def __cinit__(self):
         self.thisptr = new AhoCorasickTrie()
         self.payloads_to_decref = []
+        self.dictionary = {}
 
     def __dealloc__(self):
         for payload in self.payloads_to_decref:
@@ -123,6 +125,21 @@ cdef class NoAho:
             void_payload = <void*>py_payload
 
         self.thisptr.add_string(ucs4_data, num_ucs4_chars, void_payload)
+        self.dictionary[text] = py_payload
+
+    def save(self, fname):
+        import cPickle as pickle
+        pickle.dump(self.dictionary, open(fname, 'wb'), protocol=2)
+
+    @classmethod
+    def load(cls, fname):
+        import cPickle as pickle
+        noaho = NoAho()
+        noaho.dictionary = pickle.load(open(fname, 'rb'))
+        for key, value in noaho.dictionary.iteritems():
+            noaho.add(key, value)
+        noaho.compile()
+        return noaho
 
 # Nice-to-have...
 #    def add_many(self):
```

I believe there's something better than this.

**Comment by pquentin** [comment]
For what it's worth, this makes NoAho harder to use on Spark. It looks like the Spark driver will send an empty NoAho to drivers, and they will never report a match. A working fix is to load NoAho lazily on each worker:

```
global a
try:
    return a
except:
    a = NoAho()
    a.add(...)
    return a
```

**Comment by JDonner** [comment]
I don't think the pickling solution I've seen (an embedded dict) saves time, NoAho still has to build its tree from the dict. All I can think of is to save a tree representation and rebuild that, and hope it's faster than you could obtain adding items one-by-one (not sure it would be though, have to think about it). This could be more compact than building one-by-one, because we can know ahead of time how many slots are needed at each level of the tree. Perhaps slower, too, though. But I see that the complaints are about ease of objects-on-the-wire.

**Comment by pquentin** [comment]
Indeed, this is more about convenience right now and I just thought I would let you know. :) I'm not sure everything I'm saying is true because I only have limited experience with NoAho and (Py)Spark.

**Comment by pquentin** [comment]
(In fact, what could be nice is to make sure NoAho _cannot_ be pickled, because for now it "fails" silently.)

**Comment by pombredanne** [comment]
NoAho cannot be pickled and errors out nicely if you try..... but since the build time is rather reasonable I do not see this as an issue.
